src/coreclr/jit/assertionprop.cpp

```diff
@@ -4857,94 +4860,129 @@ bool Compiler::optNonNullAssertionProp_Ind(ASSERT_VALARG_TP assertions, GenTree*
 }
 
 //------------------------------------------------------------------------
-// optAssertionIsNonHeap: see if we can prove a tree's value will be not GC-tracked
-//   based on assertions
+// GetWriteBarrierForm: Determinate the exact type of write barrier required for the
+//    given address.
 //
 // Arguments:
-//   op         - tree to check
-//   assertions - set of live assertions
-//   pVnBased   - [out] set to true if value numbers were used
-//   pIndex     - [out] the assertion used in the proof
+//    vnStore - ValueNumStore object
+//    vn      - VN of the address
 //
 // Return Value:
-//   true if the tree's value will be not GC-tracked
+//    Exact type of write barrier required for the given address.
 //
-// Notes:
-//   Sets "pVnBased" if the assertion is value number based. If no matching
-//    assertions are found from the table, then returns "NO_ASSERTION_INDEX."
-//
-//   If both VN and assertion table yield a matching assertion, "pVnBased"
-//   is only set and the return value is "NO_ASSERTION_INDEX."
-//
-bool Compiler::optAssertionIsNonHeap(GenTree*         op,
-                                     ASSERT_VALARG_TP assertions DEBUGARG(bool* pVnBased)
-                                         DEBUGARG(AssertionIndex* pIndex))
+static GCInfo::WriteBarrierForm GetWriteBarrierForm(ValueNumStore* vnStore, ValueNum vn)
 {
-    bool vnBased = (!optLocalAssertionProp && vnStore->IsVNObjHandle(op->gtVNPair.GetConservative()));
-#ifdef DEBUG
-    *pIndex   = NO_ASSERTION_INDEX;
-    *pVnBased = vnBased;
-#endif
-
-    if (vnBased)
+    const var_types type = vnStore->TypeOfVN(vn);
+    if (type == TYP_REF)
     {
-        return true;
+        return GCInfo::WriteBarrierForm::WBF_BarrierUnchecked;
+    }
+    if (type != TYP_BYREF)
+    {
+        return GCInfo::WriteBarrierForm::WBF_BarrierUnknown;
     }
 
-    if (op->IsIconHandle(GTF_ICON_OBJ_HDL))
+    VNFuncApp funcApp;
+    if (vnStore->GetVNFunc(vnStore->VNNormalValue(vn), &funcApp))
     {
-        return true;
+        if (funcApp.m_func == VNF_PtrToArrElem)
+        {
+            // Arrays are always on the heap
+            return GCInfo::WriteBarrierForm::WBF_BarrierUnchecked;
+        }
+        if (funcApp.m_func == VNF_PtrToLoc)
+        {
+            // Pointer to a local
+            return GCInfo::WriteBarrierForm::WBF_NoBarrier;
+        }
+        if (funcApp.m_func == VNFunc(GT_ADD))
+        {
+            // Check arguments of the byref add
+            const GCInfo::WriteBarrierForm arg0Type = GetWriteBarrierForm(vnStore, funcApp.m_args[0]);
```

Comment:
better name suggestion `arg0Barrier`

```suggestion
            const GCInfo::WriteBarrierForm arg0Barrier = GetWriteBarrierForm(vnStore, funcApp.m_args[0]);
```

Reply:
Addressed